### DISCUSSION TITLE
How to add a new package on hercules

### DISCUSSION THREAD

**Discussion by DeniseWorthen** [JCSDA/spack-stack#944] [General]
I asked Hercules sys-admins to install ferret (currently on most other RDHPCS platforms as a module) but they said 
"Rocky 9 is the operating system for hercules. Ferret does not have a release for Rocky 9. We're looking into if installation is possible." 

There is a spack package though (https://spack.readthedocs.io/en/v0.16.2/package_list.html#ferret) but I'm not sure of the steps. 

I have previously added a "test package" to a personal spack-stack, but that is replacing something already there w/ something new. This is adding something that isn't already there. 

I tried the loading the modules etc for Hercules shown on 'pre-configured sites', and then tried ``spack install ferret@7.6`` but that gave me ``==> Error: Package ufs-weather-model-env does not depend on ferret@7.6.0``

**Comment by climbfuji** [comment]
Try `spack install --add ferret` - but always make sure you are in an active environment

**Comment by DeniseWorthen** [comment]
This is what I tried, but I may be doing things I shouldn't:
```
git clone --recurse-submodules https://github.com/jcsda/spack-stack -b release/1.5.1
cd spack-stack
spack stack create env --name add-ferret --template empty --site hercules --upstream /work/noaa/epic/role-epic/spack-stack/hercules/spack-stack-1.5.1/envs/unified-env/install
cd envs/add-ferret
spack env activate .
spack install --add ferret
```
Some stuff happens and then I see

```
==> Installing ferret-7.6.0-rke7wghdtkkamo27c4ddgahligozscj6
==> No binary for ferret-7.6.0-rke7wghdtkkamo27c4ddgahligozscj6 found: installing from source
==> Fetching https://mirror.spack.io/_source-cache/archive/69/69832d740bd44c9eadd198a5de4d96c4c01ae90ae28c2c3414c1bb9f43e475d1.tar.gz
==> Ran patch() for ferret
==> ferret: Executing phase: 'install'
==> Error: ProcessError: Command exited with status 2:
    'make' '-j1'

5 errors found in build log:
     112
     113    returning to directory /work/noaa/nems/dworthen/spack-stack/cache/build_stage/spack-stage-ferret-7.6.0-rke7wghdtkkamo27c4ddgahligozscj6/spack-src/xgks/port
     114
     115    make[3]: Leaving directory '/work/noaa/nems/dworthen/spack-stack/cache/build_stage/spack-stage-ferret-7.6.0-rke7wghdtkkamo27c4ddgahligozscj6/spack-src/xgks/port'
     116    make[3]: Entering directory '/work/noaa/nems/dworthen/spack-stack/cache/build_stage/spack-stage-ferret-7.6.0-rke7wghdtkkamo27c4ddgahligozscj6/spack-src/xgks/port'
     117    gcc -O -c -I. -DNDEBUG -O2 -m64 uddummy.c
  >> 118    make[3]: gcc: Permission denied
  >> 119    make[3]: *** [Makefile:193: uddummy.o] Error 127
     120    make[3]: Leaving directory '/work/noaa/nems/dworthen/spack-stack/cache/build_stage/spack-stage-ferret-7.6.0-rke7wghdtkkamo27c4ddgahligozscj6/spack-src/xgks/port'
  >> 121    make[2]: *** [Makefile:35: all] Error 2
     122    make[2]: Leaving directory '/work/noaa/nems/dworthen/spack-stack/cache/build_stage/spack-stage-ferret-7.6.0-rke7wghdtkkamo27c4ddgahligozscj6/spack-src/xgks/port'
  >> 123    make[1]: *** [Makefile:67: port/all] Error 1
     124    make[1]: Leaving directory '/work/noaa/nems/dworthen/spack-stack/cache/build_stage/spack-stage-ferret-7.6.0-rke7wghdtkkamo27c4ddgahligozscj6/spack-src/xgks'
  >> 125    make: *** [Makefile:24: optimized] Error 2

See build log for details:
  /work/noaa/nems/dworthen/spack-stack/cache/build_stage/spack-stage-ferret-7.6.0-rke7wghdtkkamo27c4ddgahligozscj6/spack-build-out.txt
```

**Comment by DeniseWorthen** [comment]
I made a second try doing this instead:

```
hercules:~> module purge
hercules:~> module use /work/noaa/epic/role-epic/spack-stack/hercules/modulefiles
hercules:~> module load ecflow/5.8.4
hercules:~> module load mysql/8.0.31
hercules:~> git clone -c feature.manyFiles=true https://github.com/spack/spack.git
hercules:~> cd spack
hercules:~> . spack/share/spack/setup-env.sh
hercules:~> spack install ferret@7.6
```

I appears to be building, so I think it is working. But I *think* I'm having to build a bunch of stuff that is already in the spack-stack (like cmake).

**Comment by DeniseWorthen** [comment]
So it finally errored out when it finally got to the ferret build. Back to the drawing board.

**Reply by climbfuji** [reply]
So that might be a problem with the ferret build system or the way it's configured in spack (mainline, not even our fork)

**Comment by DeniseWorthen** [comment]
I was able to build ferret for my own use on Hercules and I'll note here for reference. Starting from my "second try" above, I modified the compiler flag using

``spack config edit compilers``

then setting (the 'boz' might not be required, but we seem to use these in pairs in UWM CMakeLists):

```
    flags:
      fflags: -fallow-argument-mismatch -fallow-invalid-boz
```

I got a successful build, and then had to figure out *where* it was built, so I used ``spack find --paths ferret``. That gave a very long path, but from then on, I was familiar enough w/ Ferret to be able to get it set up. 

I'll note this is "classic" Ferret, not the python version (pyferret). There is no package for pyferret.

**Reply by climbfuji** [reply]
Thanks @DeniseWorthen ! I think we should patch the ferret package in spack so that users don't have to set those flags.
